pkg/storegateway/bucket.go

```diff
@@ -240,17 +250,17 @@ func NewBucketStore(
 		option(s)
 	}
 
-	snapConfig := indexheader.SnapshotterConfig{
-		Path:   dir,
-		UserID: userID,
-	}
-	s.snapshotter = indexheader.NewSnapshotter(s.logger, snapConfig)
-
-	var lazyLoadedBlocks map[ulid.ULID]int64
 	if bucketStoreConfig.IndexHeader.EagerLoadingStartupEnabled {
-		lazyLoadedBlocks = s.snapshotter.RestoreLoadedBlocks()
+		snapConfig := indexheader.SnapshotterConfig{
+			Path:            dir,
+			UserID:          userID,
+			PersistInterval: bucketStoreConfig.IndexHeader.LazyLoadingIdleTimeout,
+		}
+		s.snapshotter = indexheader.NewSnapshotter(s.logger, snapConfig, s.indexReaderPool)
+	} else {
+		s.snapshotter = noopShapshotter{services.NewIdleService(nil, nil)}
 	}
-	s.indexReaderPool = indexheader.NewReaderPool(s.logger, bucketStoreConfig.IndexHeader, s.lazyLoadingGate, metrics.indexHeaderReaderMetrics, lazyLoadedBlocks)
+	s.indexReaderPool = indexheader.NewReaderPool(s.logger, bucketStoreConfig.IndexHeader, s.lazyLoadingGate, metrics.indexHeaderReaderMetrics, s.snapshotter.RestoreLoadedBlocks())
```

Comment:
Oh, now I recall this "chicken and an egg", where the `Snapshotter` takes `IndexReader`, while the `IndexReader` needs `Snapshotter`.

Looking at it now, the `Snapshotter.RestoreLoadedBlocks` doesn't have to be a method — I made it so mainly to be symmetric to its `Snapshotter.PersistLoadedBlocks`. How about we make the former a function instead of a (static) method and reorder these constructor calls here to unwind it? This will make it less surprising for a reader.

Reply:
i actually found a bug with the current initialization - we pass a `nil` to `indexheader.NewSnapshotter`. It's fixed in the later PR, but it's still a bug here :(

I think I did what you suggested.